### PR TITLE
Feat/cache fork lookups

### PR DESCRIPTION
**Pull request by 0xgregthedev**

### CODE REVIEW

File: crates/evm/core/src/fork/backend.rs [hunk 1]
```diff
@@ -141,8 +142,10 @@ where
                 trace!(target: "backendhandler", "received request basic address={:?}", addr);
                 let acc = self.db.accounts().read().get(&addr).cloned();
                 if let Some(basic) = acc {
+                    println!("basic found for address: {:?}", addr);
```
Comment (odyslam) [inline crates/evm/core/src/fork/backend.rs:145]: We should remove print statements

File: crates/evm/core/src/fork/backend.rs [hunk 2]
```diff
@@ -141,8 +142,10 @@ where
                 trace!(target: "backendhandler", "received request basic address={:?}", addr);
```
Comment (odyslam) [inline crates/evm/core/src/fork/backend.rs:142]: better to have addr as named field

File: crates/evm/core/src/fork/init.rs [hunk 3]
```diff
@@ -1,51 +1,69 @@
...
-            if block_number <= latest_block {
-                error!("{NON_ARCHIVE_NODE_WARNING}");
-            }
-            eyre::bail!(
-                "Failed to get block for block number: {}\nlatest block number: {}",
-                block_number,
-                latest_block
-            );
+    } else if let Ok(latest_block) = provider.get_block_number().await {
+        // If the `eth_getBlockByNumber` call succeeds, but returns null instead of
```
Comment (odyslam) [inline crates/evm/core/src/fork/init.rs:55]: Nice

File: crates/evm/core/src/backend/mod.rs [hunk 4]
```diff
@@ -2005,78 +1992,84 @@ impl Backend {
...
-                match self.forks.get_fork(ForkId::new(create_fork.url.as_str(), Some(block_num))) {
+        if let Some(err) = results.find_any(|v| v.is_err()) {
+            return err;
+        }
+
+        Ok(())
+    }
+
+    /// Executes the given access on the database
+    pub fn execute_access(
```
Comment (odyslam) [inline crates/evm/core/src/backend/mod.rs:2014]: Should we add some unit tests here?

File: crates/evm/core/src/backend/environment_cache.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
+use alloy_provider::{Network, Provider};
+use alloy_rpc_types::{Block, BlockNumberOrTag};
+use alloy_transport::Transport;
+use quick_cache::sync::Cache;
+use dashmap::DashMap;
+
+#[derive(Debug)]
+pub struct EnvironmentCache {
```
Comment (odyslam) [inline crates/evm/core/src/backend/environment_cache.rs:8]: We should add some unit tests here

**Review by odyslam — APPROVED** [review]
Some nits and I feel we need to add a few more unit tests for the added functionality

File: crates/evm/core/src/fork/backend.rs (hunk 1 above)
Comment (0xgregthedev) [reply crates/evm/core/src/fork/backend.rs:145]: Nice catch, missed those.

File: crates/evm/core/src/fork/backend.rs (hunk 2 above)
Comment (0xgregthedev) [reply crates/evm/core/src/fork/backend.rs:142]: Yes, but that's not a part of the diff or relevant to the pr. 
Lots of improvements to tracing and error handling that could be made throughout the fork.

**Pull request merged**